### PR TITLE
Add BPE support for Tarred Datasets

### PR DESCRIPTION
**Pull request by titu1994**
# Changelog
- Adds BPE support to TarredDataset
- Adds respective tests for tarred bpe support
- Refactor TarredDataset to support both cases equivalently
- Add type annotations to TarredDatasets
- Add Tarred dataset support to SpeechToTextBPE models

### CODE REVIEW

File: tests/collections/asr/test_asr_datasets.py [hunk 1]
```diff
@@ -79,3 +81,34 @@ def test_tarred_dataset(self, test_data_dir):
         for _ in ds_braceexpand:
             count += 1
         assert count == 32
+
+    @pytest.mark.unit
+    def test_tarred_bpe_dataset(self, test_data_dir):
+        batch_size = 4
+        manifest_path = os.path.abspath(os.path.join(test_data_dir, 'asr/tarred_an4/tarred_audio_manifest.json'))
+
+        featurizer = WaveformFeaturizer()
```
Comment (blisc) [inline tests/collections/asr/test_asr_datasets.py:90]: Why is this defined?

Comment (titu1994) [reply tests/collections/asr/test_asr_datasets.py:90]: Just noticed that. Will revert.

Comment (titu1994) [reply tests/collections/asr/test_asr_datasets.py]: Done

File: nemo/collections/asr/models/ctc_bpe_models.py [hunk 2]
```diff
@@ -121,26 +121,48 @@ def _setup_dataloader_from_config(self, config: Optional[Dict]):
...
-            min_duration=config.get('min_duration', None),
-            max_utts=config.get('max_utts', 0),
-            trim=config.get('trim_silence', True),
-            load_audio=config.get('load_audio', True),
-            add_misc=config.get('add_misc', False),
-        )
+        shuffle = config['shuffle']
+
+        # Instantiate tarred dataset loader or normal dataset loader
+        if 'is_tarred' in config and config.get('is_tarred', False):
```
Comment (redoctopus) [inline nemo/collections/asr/models/ctc_bpe_models.py:127]: Why not just `if config.get('is_tarred', False)`?

File: nemo/collections/asr/models/ctc_bpe_models.py [hunk 3]
```diff
@@ -121,26 +121,48 @@ def _setup_dataloader_from_config(self, config: Optional[Dict]):
...
+            dataset = TarredAudioToBPEDataset(
+                audio_tar_filepaths=config['tarred_audio_filepaths'],
+                manifest_filepath=config['manifest_filepath'],
+                tokenizer=self.tokenizer,
+                sample_rate=config['sample_rate'],
+                int_values=config.get('int_values', False),
+                augmentor=augmentor,
+                shuffle_n=config.get('shuffle_n', 50),
```
Comment (redoctopus) [inline nemo/collections/asr/models/ctc_bpe_models.py:135]: Just realized I left the arbitrarily set `shuffle_n=50` in `ctc_models.py`... Do you mind adding this as well that file as well?
And I'm not sure that 50 is a good default (sorry about that); maybe something like 64 or 128, some multiple of the likely batch size.

File: tests/collections/asr/test_asr_datasets.py [hunk 4]
```diff
@@ -13,11 +13,13 @@
 # limitations under the License.
 
 import os
+
```
Comment (redoctopus) [inline tests/collections/asr/test_asr_datasets.py:16]: Extra newline?

File: nemo/collections/asr/models/ctc_bpe_models.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/asr/models/ctc_bpe_models.py:127]: Let me try if that works (it probably does)

File: nemo/collections/asr/models/ctc_bpe_models.py (hunk 3 above)
Comment (titu1994) [reply nemo/collections/asr/models/ctc_bpe_models.py:135]: Yup sounds good.

File: tests/collections/asr/test_asr_datasets.py (hunk 4 above)
Comment (titu1994) [reply tests/collections/asr/test_asr_datasets.py:16]: Seems my local copy of black is adding this explicitly. Tests version on Ci doesnt seem to mind either way..

File: nemo/collections/asr/models/ctc_bpe_models.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/asr/models/ctc_bpe_models.py:127]: Works !

File: nemo/collections/asr/models/ctc_bpe_models.py (hunk 3 above)
Comment (titu1994) [reply nemo/collections/asr/models/ctc_bpe_models.py]: Done, im defaulting to 4x batch size as a reasonable buffer for shuffle

File: tests/collections/asr/test_asr_datasets.py (hunk 4 above)
Comment (redoctopus) [reply tests/collections/asr/test_asr_datasets.py:16]: Huh, that's interesting! Well, if the style scripts decree it then I have no objections :)

**Pull request merged**